lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,416 @@
+                                    RuleNotFound, RuleReplaceFailed, UnsupportedOperation)
+from rucio.core.lock import get_replica_locks, get_replica_locks_for_rule_id
+from rucio.core.replica import list_replicas_per_rse, set_tombstone, update_replica_state
+from rucio.core.request import get_request_by_did
+from rucio.core.rse import add_rse_attribute, update_rse
+from rucio.core.rule import list_rules_for_rse_decommissioning, move_rule, update_rule
+from rucio.db.sqla.constants import ReplicaState
+
+from ..profile import DecommissioningProfile
+
+
+def generic_delete(rse: dict[str, Any], config: dict[str, Any]) -> DecommissioningProfile:
+    """Return a profile for simply deleting rules.
+
+    The "generic delete" profile lists out all rules that are locking replicas at the given RSE,
+    and deletes them if either one of the following is true:
+
+    - The RSE expression of the rule is trivial (the RSE name itself).
+    - There are no replicas locked by the rule that reside on another RSE.
+
+    :param rse: RSE to decommission.
+    :param config: Decommissioning configuration dictionary.
+    :returns: A decommissioning profile.
+    """
+    return DecommissioningProfile(
+        rse=rse,
+        initializer=_set_availability,
+        discoverer=_generic_discover,
+        handlers=[
+            (_is_being_deleted, _count_as_processed),
+            (_has_trivial_rse_expression, _delete_rule),
+            (_has_last_replica_on_rse, _delete_rule),
+        ],
+        finalizer=_generic_finalize
+    )
+
+
+def generic_move(rse: dict[str, Any], config: dict[str, Any]) -> DecommissioningProfile:
+    """Return a profile for moving rules that satisfy conditions to a specific destination.
+
+    The "generic move" profile lists out all rules that are locking replicas at the given RSE,
+    and moves them to the specified destination if either one of the following is true:
+
+    - The RSE expression of the rule is trivial (the RSE name itself).
+    - There are no replicas locked by the rule that reside on another RSE.
+
+    :param rse: RSE to decommission.
+    :param config: Decommissioning configuration dictionary.
+    :returns: A decommissioning profile dictionary.
+    """
+    try:
+        destination = config['destination']
+    except KeyError as exc:
+        raise RucioException('Configuration parameter "destination" not set') from exc
+
+    move_to_destination = RuleMover(destination)
+    return DecommissioningProfile(
+        rse=rse,
+        initializer=_set_availability,
+        discoverer=_generic_discover,
+        handlers=[
+            (_is_being_deleted, _count_as_processed),
+            (_has_child_rule_id, _count_as_processed),
+            (_has_trivial_rse_expression, move_to_destination),
+            (_has_last_replica_on_rse, move_to_destination),
+        ],
+        finalizer=_generic_finalize
+    )
+
+
+def _set_availability(
```

Comment:
This function name is misleading because it does more than just set the RSE availability. `generic_initialize` might be more apt.